Link to GitHub: https://github.com/Stampe04/CSS_Assignment1.git

All group members contributed equally to this assignment

### Part 1: Ready Made vs Custom Made Data ###

***What are pros and cons of the custom-made data used in Centola's experiment (the first study presented in the lecture) and the ready-made data used in Nicolaides's study (the second study presented in the lecture)? You can support your arguments based on the content of the lecture and the information you read in Chapter 2.3 of the book (answer in max 150 words).***

In Centola’s experiment, the custom-made data has the benefit of giving exactly what the study is looking for, without filling the data set with useless information or leaving out important information that would make the data set incomplete. However, the results might be impaired if any of the people participating were aware that they were a part of an experiment.

In Nicolaides’s experiment, the ready-made data has the benefit of being always-on, meaning that they are constantly receiving new data that can be used in their experiment. However, as it was mentioned in the video, there might be other relevant insights about the people that they do not receive from their data or that they do not have access to.

***How do you think these differences can influence the interpretation of the results in each study? (answer in max 150 words)***

Centola's study can potentially help draw strong causal conclusions, if behaviour is significantly different between the 2 network structures, because the network structures are experimentally manipulated. However, this setup may not generalize well to real world social systems, where networks are influenced by many additional factors.

Nicolaides' study can potentially create results that better reflect real world behaviour, since it uses naturally collected and large scale data. However, since the researchers cannot control for all variables, the results could only help draw correlational conclusions rather than causal.

Comparing the two, Centola's results can better show how and why behaviour spreads, while Nicolaides' results can better show how patterns appear in the real world.

### Part 2: Find Researchers using the OpenAlex API ###

We import the necessary libraries, load the dataset from the csv file, and detect which column contains author names: 

In [ ]:
import requests
import time
import pandas as pd
import re
from google.colab import drive

drive.mount('/content/drive')

data = pd.read_csv('/content/drive/MyDrive/Computational Social Science/ic2s2_2025_schedule_v5.csv')

# Find author column
author_cols = [c for c in data.columns if "author" in c.lower()]
print("Author-like columns:", author_cols)

author_col = author_cols[0] if author_cols else None
assert author_col is not None, "No author-like column found."

Mounted at /content/drive
Author-like columns: ['author']


Since the raw author cells contain inconsistent formatting like different seperators, we split each cell into individual author names, normalize the names, and deduplicate the list of authors to get a list of only unique names: 

In [ ]:
def split_authors(x):
    if pd.isna(x):
        return []
    s = str(x).strip()

    # normalize common separators
    s = s.replace("\n", ";")
    s = re.sub(r"\s+(and|&)\s+", ";", s, flags=re.IGNORECASE)
    s = s.replace("|", ";")

    parts = [p.strip() for p in s.split(";") if p.strip()]

    # If still a single blob, try to detect comma-separated FULL NAMES:
    # Example: "Étienne Ollion, Émilien Schultz"
    if len(parts) == 1 and ", " in parts[0]:
        chunks = [c.strip() for c in parts[0].split(",") if c.strip()]

        looks_like_full_names = sum((" " in c) for c in chunks) >= max(2, int(0.7 * len(chunks)))

        if looks_like_full_names:
            parts = chunks

    return parts

def clean_author_name(a):
    a = re.sub(r"\s+", " ", str(a)).strip()
    a = a.strip(" '\"")
    a = a.replace("'", "").replace('"', "")
    return a.strip()

all_authors = []
for cell in data[author_col]:
    all_authors.extend(split_authors(cell))

authors_list = []
seen = set()
for a in all_authors:
    a = clean_author_name(a)
    if a and a not in seen:
        seen.add(a)
        authors_list.append(a)

print("Number of unique authors:", len(authors_list))
print(authors_list[:30])

Number of unique authors: 1565
['Étienne Ollion', 'Émilien Schultz', 'Miriam Schirmer', 'Julia Mendelsohn', 'Dustin Wright', 'Dietram A. Scheufele', 'Ágnes Horvát', 'Caspar van Lissa', 'Jorge Barreras', 'Thomas Li', 'Chen Zhong', 'Cate Heine', 'Adam (Zhengzi) Zhou', 'Paolo Turrini', 'Elias Fernández Domingos', 'Kristina Gligoric', 'Cinoo Lee', 'Tijana Zrnic', 'Adel Daoud', 'Connor Jerzak', 'Mark Whiting', 'Linnea Gandhi', 'Amirhossein Nakhaei', 'Duncan Watts', 'Egor Kotov', 'Johannes Mast', 'Matthew A. Turner', 'James Holland Jones', 'Dean Eckles', 'Kathleen Carly']


Here we define the base URL for the OpenAlex API and a helper function that:

- sends a request with parameters
- waits between requests to avoid rate limiting
- retries if the API returns non-200 responses or error JSON
- returns a parsed JSON file when all is successful

This keeps the data collection loop we do later much cleaner.

OpenAlex returns results in pages. So we build a function that uses cursor based pagination to iterate through all the available results up to a given page limit. We use this later on when we use the works endpoint so we can collect complete datasets beyond the first page.

We also build a function that uses the OpenAlex endpoint filter to search for given concepts, which we will use later on in our filters:

In [ ]:
# Replace with your OpenAlex API key, or remove and use mailto param instead
api_key = "JW3vNYrvustrbEkExjEvWH"
BASE = "https://api.openalex.org"

# OpenAlex helper for data retrieval
def oa_get(path, params=None, sleep=0.6, tries=3):
    url = f"{BASE}{path}"
    params = dict(params or {})
    params["api_key"] = api_key  # or remove this line and use "mailto" instead if you don't have an API key

    last_err = None

    for _ in range(tries):
        try:
            r = requests.get(url, params=params, timeout=60)
            time.sleep(sleep)
            #print("Status:", r.status_code, "URL:", r.url)

            # if non-200, print body and retry
            if r.status_code != 200:
                print("Non-200 response:", r.text[:500])
                last_err = f"HTTP {r.status_code}"
                time.sleep(1.0)
                continue

            js = r.json()

            if isinstance(js, dict) and "error" in js:
                print("OpenAlex error JSON:", js)
                last_err = js
                time.sleep(1.0)
                continue

            return js

        except Exception as e:
            print("Exception in oa_get:", e)
            last_err = e
            time.sleep(1.0)

    print("oa_get failed, last_err:", last_err)
    return None

# OpenAlex helper for reading through multiple pages
def paginate(path, params=None, per_page=200, max_pages=None):
    """Generator over all results using cursor pagination."""
    p = dict(params or {})
    p["per-page"] = per_page
    p["cursor"] = "*"

    page = 0
    while True:
        js = oa_get(path, p)
        if js is None:
            break
        results = js.get("results", [])
        for item in results:
            yield item

        next_cursor = js.get("meta", {}).get("next_cursor")
        page += 1

        if not next_cursor:
            break
        if max_pages is not None and page >= max_pages:
            break

        p["cursor"] = next_cursor

# Takes the concept name (like Sociology and Computer Science) and returns a concept ID
def get_concept_id_by_name(name):
    js = oa_get("/concepts", {
        "search": name,
        "select": "id,display_name,level,works_count",
        "per-page": 10
    })
    results = js.get("results", [])
    if not results:
        return None

    name_l = name.strip().lower()
    for c in results:
        if (c.get("display_name") or "").strip().lower() == name_l:
            return c.get("id")

    return results[0].get("id")

This is the function we use to find researchers from the schedule. This is done by using the function mentioned above. The information we gather is as requested:

- author_id
- display_name
- works_count
- h_index
- country_code

Some of these informations are not directly searchable, but requries searching deeper in certain lists or dictionaries:

In [ ]:
def find_author_tophit(author_name):
    js = oa_get("/authors", {
        "search": author_name,
        "select": "id,display_name,works_count,summary_stats,last_known_institutions",
        "per-page": 1
    })
    if js is None:
        return None

    results = js.get("results", [])
    if not results:
        return None

    r = results[0]
    insts = r.get("last_known_institutions") or []
    country_code = insts[0].get("country_code") if insts else None

    return {
        "id": r.get("id"),
        "display_name": r.get("display_name"),
        "works_count": r.get("works_count"),
        "h_index": (r.get("summary_stats") or {}).get("h_index"),
        "country_code": country_code,
        "query_name": author_name,
    }

Here we first retrieve OpenAlex concept IDs for the selected social science and quantitative fields. Next, we clean and normalize the author names from the dataset to make matching more reliable.

We then paginate through the OpenAlex authors endpoint (restricted to authors with 5–5000 works) and keep only authors whose display names match the schedule.

For each match we also store basic metadata (id, name, works count, h-index, and country), which creates dataset D1.  The matched author IDs are saved as `eligible_authors`, which we be use later in Part 3.

In [ ]:
# 1. Get concept IDs
social_names = ["Sociology", "Psychology", "Economics", "Political science"]
quant_names  = ["Mathematics", "Physics", "Computer science"]

social_ids = [get_concept_id_by_name(n) for n in social_names]
quant_ids  = [get_concept_id_by_name(n) for n in quant_names]
social_ids = [x for x in social_ids if x]
quant_ids  = [x for x in quant_ids if x]

assert len(social_ids) > 0, "Could not find any social concept IDs."
assert len(quant_ids)  > 0, "Could not find any quantitative concept IDs."

concept_or = "|".join(social_ids + quant_ids)

# Normalize authors_list for matching (lowercase, strip and commas)
authors_set = set()
for a in authors_list:
    if not a:
        continue
    cleaned = clean_author_name(a).lower()
    authors_set.add(cleaned)
    # OpenAlex often uses "Last, First" - add "first last" as variant for matching
    if ", " in cleaned:
        parts = [p.strip() for p in cleaned.split(",", 1)]
        if len(parts) == 2:
            authors_set.add(f"{parts[1]} {parts[0]}")

# Paginate /authors filtered by works_count
D1 = []
MAX_AUTHOR_PAGES = 200  # pages to search (200 authors/page); None = no limit

params = {
    "filter": "works_count:>4,works_count:<5001",
    "select": "id,display_name,works_count,summary_stats,last_known_institutions",
    "sort": "works_count:desc",
}

def name_matches(dn, schedule_set):
    #Check if OpenAlex display_name matches any author
    dn_lower = dn.lower().strip()
    if dn_lower in schedule_set:
        return True
    # OpenAlex "Last, First" -> try "first last"
    if ", " in dn_lower:
        parts = [p.strip() for p in dn_lower.split(",", 1)]
        if len(parts) == 2:
            if f"{parts[1]} {parts[0]}" in schedule_set:
                return True
    return False

matched_names = set()  # track which authors we've found
for a in paginate("/authors", params, per_page=200, max_pages=MAX_AUTHOR_PAGES):
    display_name = (a.get("display_name") or "").strip()
    if not display_name:
        continue
    if name_matches(display_name, authors_set):
        insts = a.get("last_known_institutions") or []
        country_code = insts[0].get("country_code") if insts else None
        D1.append({
            "id": a.get("id"),
            "display_name": display_name,
            "works_count": a.get("works_count"),
            "h_index": (a.get("summary_stats") or {}).get("h_index"),
            "country_code": country_code,
        })
        # Track matched name
        dn_lower = display_name.lower()
        matched_names.add(dn_lower)
        if ", " in dn_lower:
            parts = [p.strip() for p in dn_lower.split(",", 1)]
            if len(parts) == 2:
                matched_names.add(f"{parts[1]} {parts[0]}")
    # Early exit: stop when we've found all authors
    if matched_names >= authors_set:
        print("Found all schedule authors, stopping early.")
        break

df_auth = pd.DataFrame(D1)
display(df_auth.head(10))

# For Part 3
eligible_authors = set(df_auth["id"].dropna().tolist())
eligible_df = df_auth.copy()
print("Eligible authors:", len(eligible_authors))

,id,display_name,works_count,h_index,country_code
0,https://openalex.org/A5078999329,Young Jin Kim,3732,8,KR
1,https://openalex.org/A5018012012,Gang Li,3243,44,CN
2,https://openalex.org/A5059645286,Robert West,2617,132,GB
3,https://openalex.org/A5100329996,Xi Chen,2257,114,CN
4,https://openalex.org/A5100322712,Yan Wang,1962,75,CN
5,https://openalex.org/A5100438687,Gang Li,1781,92,CN
6,https://openalex.org/A5101619735,Andreas Maier,1698,53,DE
7,https://openalex.org/A5089411227,Hong Chen,1577,73,None
8,https://openalex.org/A5055292867,Peng Liu,1362,8,CN
9,https://openalex.org/A5100438696,Gang Li,1333,81,US


Eligible authors: 65


***Which challenges did you encounter? How did you address them?***

In the beginning, we had some issues with the API itself, since we found the API documentation very confusing to navigate. However, this slowly became less of an issue as we went along and got more familiar with the documentation and the code. Furthermore, we had some initial issues loading the csv file properly, but this was solved by doing some simple research on handling csv files in Python. In addition, we ran into trouble with our API keys multiple times, as we kept running into issues with having enough tokens per day. We solved this by testing our code on a limited amount of data to use our keys sparingly, and then do the full run once we knew the code worked.

***Choose one problem you faced while collecting the data and describe your solution. Why did you choose this approach, and what impact might it have on your data?***

The biggest problem would be the limitation issues with our API keys, as we tested whole tables during our coding trials which put quite a strain on our amount of token usage. We solved this problem by limiting the number of calls with our API keys, e.g. limiting the number of data we called to ensure the code worked for all rows of data. We chose this approach as we still needed to test our code for edge cases and this would do just that without limiting our testing to a few times a day.

Another problem we faced was loading the .csv file from the IC2S2 website using web scraping. Due to the fact that the website's dataset was transformed whenever we web scraped, we needed to locate the dataset and extract it before web scraping in Python. This wasn't the biggest issue as we could extract the HTML information from the website and locate it, but it proved to be a hassle to find it without downloading all dependencies (i.e.  the whole website).

### Part 3: Collect Research Articles ###

We started of by creating OR-strings from the social science and quantitative concept IDs found earlier.  

These will be used in the works query to filter papers so that each result belongs to at least one social science concept and one quantitative concept, which was required in Part 3.

In [ ]:
# Concept IDs for works filter (already computed above; define social_or, quant_or)
social_or = "|".join(social_ids)
quant_or  = "|".join(quant_ids)
print("Social concept OR:", social_or)
print("Quant concept OR:", quant_or)

Social concept OR: https://openalex.org/C144024400|https://openalex.org/C15744967|https://openalex.org/C162324750|https://openalex.org/C17744445
Quant concept OR: https://openalex.org/C33923547|https://openalex.org/C121332964|https://openalex.org/C41008148


Here we define a helper to extract all co-author OpenAlex IDs from a work. We then query the `/works` endpoint for our eligible authors, using the requested API-side filters (citations > 10, < 10 authors, and at least one social + one quantitative concept).

To avoid URL-length errors, we split the author IDs into chunks and run the works query per chunk.  

We also made sure to deduplicate by work ID, and store paper metadata in D2 and title/abstract data in D3.

In [ ]:
def extract_author_ids(work):
    auths = work.get("authorships") or []
    ids = []
    for a in auths:
        au = (a.get("author") or {}).get("id")
        if au:
            ids.append(au)
    return ids

# Build OR filter over all eligible authors
eligible_ids = [aid for aid in eligible_authors if aid]
print("Number of eligible author IDs in filter:", len(eligible_ids))

# Chunk author IDs to avoid "Request Line is too large" (URL limit ~4KB)
CHUNK_SIZE = 25
D2_rows, D3_rows = [], []
seen_work_ids = set()
n_seen = 0
MAX_PAGES = None  # e.g. set to 2 for quick tests

for i in range(0, len(eligible_ids), CHUNK_SIZE):
    chunk = eligible_ids[i:i + CHUNK_SIZE]
    author_or = "|".join(chunk)

    filter_str = (
        f"authorships.author.id:{author_or},"
        f"cited_by_count:>10,"
        f"authors_count:<10,"
        f"concept.id:{social_or},"
        f"concept.id:{quant_or}"
    )

    params = {
        "filter": filter_str,
        "select": "id,publication_year,cited_by_count,authorships,title,abstract_inverted_index",
        "sort": "cited_by_count:desc",
    }

    for w in paginate("/works", params, per_page=200, max_pages=MAX_PAGES):
        n_seen += 1
        wid = w.get("id")
        if not wid or wid in seen_work_ids:
            continue

        author_ids = extract_author_ids(w)

        D2_rows.append({
            "id": wid,
            "publication_year": w.get("publication_year"),
            "cited_by_count": w.get("cited_by_count"),
            "author_ids": author_ids,
        })

        D3_rows.append({
            "id": wid,
            "title": w.get("title"),
            "abstract_inverted_index": w.get("abstract_inverted_index"),
        })

        seen_work_ids.add(wid)

    print("Works returned (including filtered-out duplicates):", n_seen)
    print("Unique works kept:", len(seen_work_ids))

    df2 = pd.DataFrame(D2_rows)
    df3 = pd.DataFrame(D3_rows)

print("D2 shape:", df2.shape)
print("D3 shape:", df3.shape)

Number of eligible author IDs in filter: 65
Works returned (including filtered-out duplicates): 884
Unique works kept: 884
Works returned (including filtered-out duplicates): 2225
Unique works kept: 2225
Works returned (including filtered-out duplicates): 2929
Unique works kept: 2928
D2 shape: (2928, 4)
D3 shape: (2928, 3)


Here we verify that neither D2 or D3 contains duplicates, and that both datasets cover the exact same set of works.

We also display the first few elements of each dataset to visualize some of the results:

In [ ]:
display(df2.head(20))
display(df3.head(20))

print("Duplicate IDs in D2?", df2["id"].duplicated().any())
print("Duplicate IDs in D3?", df3["id"].duplicated().any())
print("Same ID sets?", set(df2["id"]) == set(df3["id"]))

,id,publication_year,cited_by_count,author_ids
0,https://openalex.org/W744855463,2014,2371,"[https://openalex.org/A5090631705, https://ope..."
1,https://openalex.org/W2953356739,2019,1376,"[https://openalex.org/A5101584823, https://ope..."
2,https://openalex.org/W2158437999,2001,1268,[https://openalex.org/A5063647785]
3,https://openalex.org/W2279053323,2015,1195,"[https://openalex.org/A5050740290, https://ope..."
4,https://openalex.org/W2515462165,2016,1093,"[https://openalex.org/A5043098453, https://ope..."
5,https://openalex.org/W91442942,2011,1013,"[https://openalex.org/A5007374520, https://ope..."
6,https://openalex.org/W2084938591,2013,1000,"[https://openalex.org/A5009970175, https://ope..."
7,https://openalex.org/W2242161203,2015,893,"[https://openalex.org/A5060417049, https://ope..."
8,https://openalex.org/W2595521492,2017,871,"[https://openalex.org/A5031461158, https://ope..."
9,https://openalex.org/W3133542152,2022,862,"[https://openalex.org/A5100700956, https://ope..."


,id,title,abstract_inverted_index
0,https://openalex.org/W744855463,The Behaviour Change Wheel: A Guide To Designi...,"{'This': [0], 'is': [1, 15, 41], 'a': [2, 22, ..."
1,https://openalex.org/W2953356739,ERNIE: Enhanced Language Representation with I...,"{'Neural': [0], 'language': [1, 36, 53, 64, 84..."
2,https://openalex.org/W2158437999,On the Limits of Framing Effects: Who Can Frame?,"{'Public': [0], 'opinion': [1], 'often': [2], ..."
3,https://openalex.org/W2279053323,The Generalizability of Survey Experiments,"{'Abstract': [0], 'Survey': [1], 'experiments'..."
4,https://openalex.org/W2515462165,Neural Relation Extraction with Selective Atte...,"{'Distant': [0], 'supervised': [1], 'relation'..."
5,https://openalex.org/W91442942,Political Polarization on Twitter,"{'This': [0], 'is': [1], 'the': [2, 8, 12, 30]..."
6,https://openalex.org/W2084938591,Erratum: Competition among memes in a world wi...,"{'The': [0, 89], 'wide': [1], 'adoption': [2],..."
7,https://openalex.org/W2242161203,Network representation learning with rich text...,"{'Representation': [0], 'learning': [1, 18, 21..."
8,https://openalex.org/W2595521492,"Online Human-Bot Interactions: Detection, Esti...","{'Increasing': [0], 'evidence': [1], 'suggests..."
9,https://openalex.org/W3133542152,Generalizing to Unseen Domains: A Survey on Do...,"{'Machine': [0], 'learning': [1, 144], 'system..."


Duplicate IDs in D2? False
Duplicate IDs in D3? False
Same ID sets? True


Finally, we calculate how many works there are in D2, and how many unique author IDs that appear across those works:

In [ ]:
# Dataset D2/D3 summary for report

# Number of works in D2
n_works = len(df2)

# Number of unique authors across all works in D2
all_author_ids = set()
for ids in df2["author_ids"]:
    for aid in ids:
        all_author_ids.add(aid)

n_authors = len(all_author_ids)

print(f"Number of works in D2: {n_works}")
print(f"Number of unique authors in D2: {n_authors}")

Number of works in D2: 2928
Number of unique authors in D2: 5588


***How many works are listed in your Dataset D2 (IC2S2 papers) dataframe? How many unique researchers have co-authored these works?***

This is also answered in the cell above. There are 2928 works in D2 and the number of unique authors that have co-authored are 5588

***Describe the strategies you implemented to make your code more efficient. How did your approach affect your code's execution time?***

We used a bulk /authors query (one paginated request filtered by concepts and works_count) instead of searching for each author by name, which drastically reduced API calls for Part 2. For Part 3, we use a single /works query filtered by eligible author IDs, concepts, citation count, and author count—instead of querying works per author. Server-side filters (cited_by_count, authorships_count, concept.id) ensure OpenAlex returns only relevant results, which speeds up the process and reduces data transfer.

***Reflect on the rationale behind setting specific thresholds for the total number of works by an author, the citation count, the number of authors per work, and the relevance of works to specific fields. How do these filtering criteria contribute to the relevance of the dataset you compiled? Do you believe any aspects of Computational Social Science research might be underrepresented or overrepresented as a result of these choices?***

Before applying any of the filters or specific settings, we ran a testing round on just one person, which alone gave us an enormous output, and it was very difficult to find specific information in it. This alone made us realize how necessary it would be to apply filters when searching for multiple people, but in practice, we also realised just how powerful of a tool it can be, especially when there is very specific information you are interested in or a specific criteria or threshold you want to limit yourself to. Overall, we think they are an excellent tool.  

However, we also see some potential problem when applying these filters. As an example, if you choose to filter based on a work or an author that have been cited at least x amount of times, you may underrepresent authors or works that are legitimately good but have just not been used much or discovered yet. Moreover, this can overrepresent works or authors that may not be good anymore or have become outdated, just because they have been cited a lot.